3-enrich/spark-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.spark/EnrichJob.scala

```diff
@@ -0,0 +1,173 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and
+ * limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich
+package spark
+
+// Apache commons
+import org.apache.commons.codec.binary.Base64
+
+// Scalaz
+import scalaz._
+import Scalaz._
+
+// Spark
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.serializer.KryoSerializer
+import org.apache.spark.sql.{Encoders, SparkSession}
+import org.apache.spark.sql.functions.col
+
+// Snowplow
+import common.{EtlPipeline, FatalEtlError, ValidatedEnrichedEvent}
+import common.loaders.{Loader, ThriftLoader}
+import common.outputs.{BadRow, EnrichedEvent}
+
+object EnrichJob extends SparkJob {
+  private[spark] val classesToRegister: Array[Class[_]] = Array(
+    classOf[Array[String]],
+    classOf[Array[Array[Byte]]],
+    classOf[EnrichedEvent],
+    classOf[scala.collection.immutable.Map$EmptyMap$],
+    classOf[scala.collection.immutable.Set$EmptySet$],
+    classOf[scala.collection.mutable.WrappedArray$ofRef],
+    classOf[org.apache.spark.internal.io.FileCommitProtocol$TaskCommitMessage]
+  )
+  override def sparkConfig(): SparkConf = new SparkConf()
+    .setAppName(getClass().getSimpleName())
+    .setIfMissing("spark.master", "local[*]")
+    .set("spark.serializer", classOf[KryoSerializer].getName())
+    .set("spark.kryo.registrationRequired", "true")
+    .registerKryoClasses(classesToRegister)
+
+  override def run(spark: SparkSession, args: Array[String]): Unit = {
+    val job = EnrichJob(spark, args)
+    job.run()
+  }
+
+  def apply(spark: SparkSession, args: Array[String]) = new EnrichJob(spark, args)
+
+  val etlVersion = s"spark-${generated.ProjectSettings.version}"
+
+  /**
+   * Project our Failures into a List of Nel of strings.
+   * @param all A List of Validations each containing either an EnrichedEvent or Failure strings
+   * @return A (possibly empty) List of failures, where each failure is a Nel of strings
+   */
+  def projectBads(all: List[ValidatedEnrichedEvent]): List[NonEmptyList[String]] =
+    all.collect { case Failure(errs) => errs }
+
+  /**
+   * Project our Sucesses into EnrichedEvents.
+   * @param all A List of Validations each containing either an EnrichedEvent or Failure strings
+   * @return A (possibly empty) List of EnrichedEvent
+   */
+  def projectGoods(all: List[ValidatedEnrichedEvent]): List[EnrichedEvent] =
+    all.collect { case Success(e) => e }
+}
+
+/**
+ * The Snowplow Enrich job, written in Spark.
+ * @param spark Spark session used throughout the job
+ * @param args Command line arguments for the enrich job
+ */
+class EnrichJob(@transient val spark: SparkSession, args: Array[String]) extends Serializable {
+  @transient private val sc: SparkContext = spark.sparkContext
+  import spark.implicits._
```

Comment:
It seems `toDF` is part of `DataSet` class. And it also compiles without this import.